### ISSUE TITLE
Limit (indirect) dependencies in the `rego` package (esp. prometheus)

### ISSUE THREAD

**Issue by oxisto** [open-policy-agent/opa#6754]
## What is the underlying problem you're trying to solve?

Hi! We are currently using the Rego Go API (not the SDK) to execute Rego files. This works quite well also with the "storage" options. There is only one drawback to our current solution. It seems that using the `rego` package introduces quite a lot of indirect dependencies, that we do not really need or use. This is mainly prometheus (including 10 indirect dependencies on prometheus itself) and open telemetry trace. In total, just importing the `rego` package imports the following packages:
* github.com/beorn7/perks
* github.com/cespare/xxhash/v2
* github.com/prometheus/client_golang
* github.com/prometheus/client_model
* github.com/prometheus/common
* github.com/prometheus/procfs
* go.opentelemetry.io/otel
* go.opentelemetry.io/otel/metric
* go.opentelemetry.io/otel/sdk
* go.opentelemetry.io/otel/trace

This seems to be in contrast to the comment made here, https://github.com/open-policy-agent/opa/issues/2348#issuecomment-621132496, which said that plugins should not depend on prometheus.

It pulls even more dependencies from the `ast` and `topdown` packages, but that is besides the point for now. (Interesting side-node.. `plugins` uses github.com/cespare/xxhash/v2 from prometheus and `last` uses github.com/OneOfOne/xxhash).

As far as I can see the "culprit" is the `pluginMgr` field in the `Rego` struct, which pulls in the `plugins` package, which in turn pulls in all of the above.

## Describe the ideal solution

Probably, `plugins` should work without prometheus and/or open telemetry, but I am not too familiar with the

Ideally, if we are not using plugins (or the plugin manager), the package `plugins` should not be pulled in 

## Describe a "Good Enough" solution

Maybe, the `PluginManager` could live somewhere else in another package, so that `plugins` is not pulled but only this (common) package between `rego` and `plugins`.

## Additional Context

Not really, but I am open to help in implementing this, once a design decision has been reached.

**Comment by oxisto** [comment]
It seems that https://github.com/open-policy-agent/opa/pull/5939 introduces this issue

**Comment by srenatus** [comment]
Yeah stuff like that's creeps in unintentionally. Ideally, some liner would protect our intentions. 

Now, we cannot really move plugin.Manager, I think, but there might be other solutions...? Similar to how we handled #6500, maybe?

**Comment by oxisto** [comment]
> Yeah stuff like that's creeps in unintentionally. Ideally, some liner would protect our intentions.
> 
> Now, we cannot really move plugin.Manager, I think, but there might be other solutions...? Similar to how we handled [#6500](https://github.com/open-policy-agent/opa/pull/6500), maybe?

This is actually still a major problem, we are using Rego in various project but the dependency creep because of otel is really insane. Any way forward? Happy to help prepare a PR if needed but last time it sounded like you need to do some additional things internally first?

**Comment by srenatus** [comment]
Heya!

Let's figure this out. The ideal would be to have a small package that embodies the parser, compiler and evaluator of Rego, so the v1/rego package (plus v1/ast and v1/topdown). We're constrained by not breaking existing code (too badly 😏).

At the same time, I wonder if we need to reach the ideal to make things _better than they are right now_. You mention OTel specifically. I've since then come across [franz-go](https://github.com/twmb/franz-go)'s approach to dependencies, which I like a lot: it features a large number of generic extension points, when allows you to plug in whatever you need. You only "pay" for the dependency burden of, say, opentelemetry, when you actually use it via [the kotel "plugin"](https://github.com/twmb/franz-go/tree/master/plugin/kotel). Each plugin is its own go module to make this work.

The question is then: How can we adjust our code (and module) layout such that
1. nothing changes for the "big builds", i.e. the binaries and docker images we distribute
2. source-level users get a real choice

I think it would be acceptable to require the (2.) user base to touch their code -- like, if they need to add an import or a line of code -- if we bring this about with caution and ample warnings. The only troublesome usage is the v1/sdk package: it embodies the same runtime as "OPA the server", with the same config. Users of the v1/rego package, for example, already pass along their distributed tracing configs if they want that feature.

What do you think? Am I missing something?

Actually, now I've written all that, I've looked into the code once more and realized two things: a) you don't care about the (transitive) dependencies, but rather about the modules being imported and activated. b) A simple interface trick might help here, I'll go try something. 🔍 

(I believe we have crossed paths when I was visiting Garching, working for AISEC/SSE some years ago 👋 I'm happy to learn that you're OPA users.)

**Comment by oxisto** [comment]
> Heya!
> 
> Let's figure this out. The ideal would be to have a small package that embodies the parser, compiler and evaluator of Rego, so the v1/rego package (plus v1/ast and v1/topdown). We're constrained by not breaking existing code (too badly 😏).
> 
> At the same time, I wonder if we need to reach the ideal to make things _better than they are right now_. You mention OTel specifically. I've since then come across [franz-go](https://github.com/twmb/franz-go)'s approach to dependencies, which I like a lot: it features a large number of generic extension points, when allows you to plug in whatever you need. You only "pay" for the dependency burden of, say, opentelemetry, when you actually use it via [the kotel "plugin"](https://github.com/twmb/franz-go/tree/master/plugin/kotel). Each plugin is its own go module to make this work.
> 
> The question is then: How can we adjust our code (and module) layout such that
> 
> 1. nothing changes for the "big builds", i.e. the binaries and docker images we distribute
> 2. source-level users get a real choice
> 
> I think it would be acceptable to require the (2.) user base to touch their code -- like, if they need to add an import or a line of code -- if we bring this about with caution and ample warnings. The only troublesome usage is the v1/sdk package: it embodies the same runtime as "OPA the server", with the same config. Users of the v1/rego package, for example, already pass along their distributed tracing configs if they want that feature.
> 
> What do you think? Am I missing something?

That solution looks very good! Yeah we are basically just using the core Rego functionality without any plugins so I think this would work very well. I will try it out with the current main

> 
> Actually, now I've written all that, I've looked into the code once more and realized two things: a) you don't care about the (transitive) dependencies, but rather about the modules being imported and activated. b) A simple interface trick might help here, I'll go try something. 🔍
> 
> (I believe we have crossed paths when I was visiting Garching, working for AISEC/SSE some years ago 👋 I'm happy to learn that you're OPA users.)

Hehe yeah I remembered your name, happy coincidences :)

**Comment by srenatus** [comment]
Hm. Given your commit there, I don't think we really urgently need a module split. If the (transitive) dependencies disappear from `go.mod` just because the packages in OPA that use them are no longer in your import tree, then that should be sufficient. 🤔 

As sone thing we could do, I think we could add some check that puts some assertions on the package tree of `v1/rego`, to avoid blowing up that part of the tree that much again :point_up:

**Comment by oxisto** [comment]
> Hm. Given your commit there, I don't think we really urgently need a module split. If the (transitive) dependencies disappear from `go.mod` just because the packages in OPA that use them are no longer in your import tree, then that should be sufficient. 🤔

Yes, Go pruning will do an awesome job as long as you are not including it in that *package*, it does not have to be on different modules.

> 
> As sone thing we could do, I think we could add some check that puts some assertions on the package tree of `v1/rego`, to avoid blowing up that part of the tree that much again ☝️

Nice!
